### PR TITLE
Start of the new execution engine implementation

### PR DESCRIPTION
**Pull request by andimarek**
wip for a rewrite of the execution engine. includes batched execution already.

### CODE REVIEW

File: src/main/java/graphql/execution/Async.java [hunk 1]
```diff
@@ -132,4 +134,43 @@
         });
     }
 
+
+    public static <U, T> CompletableFuture<U> reduce(List<CompletableFuture<T>> values, U initialValue, BiFunction<U, T, U> aggregator) {
```
Comment (bbakerman) [inline src/main/java/graphql/execution/Async.java]: javadoc??

File: src/main/java/graphql/execution/Async.java [hunk 2]
```diff
@@ -132,4 +134,43 @@
...
+        }
+        values.get(curIndex).
+                thenApply(oneValue -> aggregator.apply(curValue, oneValue))
+                .thenAccept(newValue -> reduceImpl(values, curIndex + 1, newValue, aggregator, result));
+    }
+
+    public static <U, T> CompletableFuture<List<U>> map(CompletableFuture<List<T>> values, Function<T, U> mapper) {
+        return values.thenApply(list -> list.stream().map(mapper).collect(Collectors.toList()));
+    }
+
```
Comment (bbakerman) [inline src/main/java/graphql/execution/Async.java]: andi-reactor code! ;)

File: src/main/java/graphql/execution2/BatchedDataFetcher.java [hunk 3]
```diff
@@ -0,0 +1,6 @@
+package graphql.execution2;
+
+import graphql.schema.DataFetcher;
+
+public interface BatchedDataFetcher<T> extends DataFetcher<T> {
+}
```
Comment (bbakerman) [inline src/main/java/graphql/execution2/BatchedDataFetcher.java]: javadoc eventually

File: src/main/java/graphql/execution2/BatchedExecutionStrategy.java [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.stream.Collectors;
+
+import static java.util.stream.Collectors.groupingBy;
+import static java.util.stream.Collectors.toList;
+
+public class BatchedExecutionStrategy implements ExecutionStrategy {
```
Comment (bbakerman) [inline src/main/java/graphql/execution2/BatchedExecutionStrategy.java]: javadoc on purpose eventually

File: src/main/java/graphql/execution2/BatchedExecutionStrategy.java [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+
+    ExecutionStepInfoFactory executionInfoFactory;
+    ValueFetcher valueFetcher;
+    ResultNodesCreator resultNodesCreator = new ResultNodesCreator();
+
```
Comment (bbakerman) [inline src/main/java/graphql/execution2/BatchedExecutionStrategy.java]: private scope and final??

File: src/main/java/graphql/execution2/Common.java [hunk 6]
```diff
@@ -0,0 +1,72 @@
+package graphql.execution2;
```
Comment (bbakerman) [inline src/main/java/graphql/execution2/Common.java]: Common seems like a kitchen sink class.  Is there not another home or a  better named class maybe.

File: src/main/java/graphql/execution2/Common.java [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+
+    public static <T> CompletableFuture<List<T>> flatList(CompletableFuture<List<List<T>>> cf) {
+        return cf.thenApply(Common::flatList);
+    }
+
+    public static <T> List<T> flatList(List<List<T>> listLists) {
+        return listLists.stream()
+                .flatMap(List::stream)
+                .collect(Collectors.toList());
+    }
```
Comment (bbakerman) [inline src/main/java/graphql/execution2/Common.java]: We have FpKit already for some of the methods.  I would move them there eventually

File: src/main/java/graphql/execution2/DefaultExecutionStrategy.java [hunk 8]
```diff
@@ -0,0 +1,112 @@
+package graphql.execution2;
```
Comment (bbakerman) [inline src/main/java/graphql/execution2/DefaultExecutionStrategy.java]: Why the `execution2` path.  Why not called leave it in `graphql.execution` and rename the class.

is this intended to be the packahe on going.  `graphql.xxx2` smells to me

File: src/main/java/graphql/execution2/DefaultExecutionStrategy.java [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+import java.util.concurrent.CompletableFuture;
+import java.util.stream.Collectors;
+
+import static java.util.stream.Collectors.toList;
+
+public class DefaultExecutionStrategy implements ExecutionStrategy {
+
+    ExecutionStepInfoFactory executionInfoFactory;
+    ValueFetcher valueFetcher;
+    ResultNodesCreator resultNodesCreator = new ResultNodesCreator();
```
Comment (bbakerman) [inline src/main/java/graphql/execution2/DefaultExecutionStrategy.java]: private final??

File: src/main/java/graphql/tuples/Tuple7.java [hunk 10]
```diff
@@ -0,0 +1,208 @@
+/*
```
Comment (bbakerman) [inline src/main/java/graphql/tuples/Tuple7.java]: Do we need all these Tuples classes.  They seem unused at present?

**Comment by kaqqao** [comment]
Any chance you give a summary of the these changes? I'm curious to know what's coming and it's a bit difficult to follow without knowing the goal/motivation behind the changes. How is the new default execution strategy different form the current etc...
Also, does this mean the now deprecated `BatchedExecutionStrategy` is getting rehabilitated?

**Comment by bbakerman** [comment]
@kaqqao - let me give you the 2 minute explanation and Andi can give you the long winded version since its his baby.

The thought is to create a  "next gen" execution engine that is truly async by design as opposed to todays which was created "sync" and then evolved into a CF based one.

The other goal is to allow it to be designed to easily allow "batching" by default and hence not need a "batchES" or "not batch ES"

This work came out of some explorations in a "reactive" execution strategy that then turned into a CompletableFuture based one so it can run on Java native without a hard dependency on say rxjava or reactor libs.

We think this will "co-exist" with the existing ES modules for some time.

We also need to re-think "Instrumentation" and so on since this has a completely different execution shape than the old "depth first" field at a  time pattern.

**Comment by kaqqao** [comment]
Thanks, @bbakerman! Sounds like a deep incision indeed 😮 I'll definitely have to look into it in detail. Don't want to just end up with the world suddenly changing under my feet.

**Comment by andimarek** [comment]
@kaqqao what brad said: it is rewrite of the core execution engine to allow for more powerful execution strategies which are still maintainable.

The old execution strategy will still be around for the foreseeable future, but the new one will offer a spec complaint first level batching implementation. Also things like defer or streaming or more advanced execution optimization should be possible or more easily possible compared to the current implementation.

Because there are still quite a few things to figure out this will be internal for now and will be public (or maybe experimental) once we are sure it does what it should.

Just to be clear again: the are no plans to retire the old execution engine!

File: src/main/java/graphql/execution/Async.java (hunk 2 above)
Comment (andimarek) [reply src/main/java/graphql/execution/Async.java]: yea ... tradeoffs :-)

**Comment by andimarek** [comment]
This will be merged despite not being ready: we are planning on evolving it incrementally on master.

**Comment by bbakerman** [comment]
@exbe - this is not yet callable - the code exists but is not wired in from the top.  So you can use it today from public API - only on tests so far.

But yes later we will come up with a mechanism to "use this next gen" engines

**Pull request merged**